codebase/src/compiler/checker-relations.md

```diff
@@ -0,0 +1,90 @@
+
+> **_Note:_** something that can be confusing is how `any` has a different behavior depending on the relation being used to compare it, and how it differs from `unknown`. Basically, `any` is special because it is assignable to every type, but `any` is not a subtype of every type. And every type is assignable to `any`, and every type is a subtype of `any`. So `any` has this special behavior during assignability where it is assignable to **any**thing.
+`unknown`, like `any`, is also a supertype of every other type. However, `unknown` doesn't have `any`'s special assignability behavior: everything is assignable to `unknown`, but `unknown` is not assignable to everything.
+
+Another relation we have is **comparability**. We use comparability to check if two types possibly overlap and values of those types can possibly be compared to one another. For example:
+
+```ts
+type Color = "red" | "blue" | "green" | "orange";
+type Fruit = "apple" | "banana" | "orange";
+type Shape = "square" | "rectangle";
+
+function myFoo(color: Color, fruit: Fruit, shape: Shape) {
+    color = fruit; // Error: Type 'Fruit' is not assignable to type 'Color'.
+    fruit = color; // Error: Type 'Color' is not assignable to type 'Fruit'.
+    if (color === fruit) { // Ok, because 'Fruit' and 'Color' are comparable.
+        ...
+    }
+    if (color === shape) { // Error: This comparison appears to be unintentional because the types 'Color' and 'Shape' have no overlap.
+        ...
+    }
+}
+```
+The types `Color` and `Fruit`, despite not being assignable to each other, are comparable, because if we think of types as sets, they have values in common (in this case, value `"orange"`). However, types `Color` and `Shape` are not comparable, so the checker will complain if you try to compare values of those two types.
+
+As for the **identity** relation, it means to be used as its name implies: to check if two types are **identical**. We use that one more rarely in the checker. One example is when you declare the same `var` multiple times (which is allowed), the checker checks that the declared types are identical among all of the declarations.
+
+Another usage of identity is in inference. When inferring from one type into another, we get rid of the parts of both types that are identical. For example, when inferring from one union to another union, e.g. from `boolean | string | number` to `T | string | number`, we get rid of the identical types on both sides (`string` and `number`) to end up inferring from `boolean` into `T`.
+Yet another example is when checking assignability of two instantiations of the same invariant generic type: two instantiations are still related if their type arguments is identical.
+
+> **_Implementation note:_** Most of the times, identity is used for object types, and we try to make it quick to check that by interning that information.
```

Comment:
I don't know if identity is special in this regard. We do a fast equality check and cache object types for every type relationship.